wave_matching/src/pointcloud_display.cpp

```diff
@@ -0,0 +1,117 @@
+#include "wave/matching/pointcloud_display.hpp"
+#include "wave/utils/logging.hpp"
+
+namespace wave {
+
+PointcloudDisplay::PointcloudDisplay() {}
+
+void PointcloudDisplay::startSpin() {
+    this->continueFlag.test_and_set(std::memory_order_relaxed);
+    this->viewer_thread =
+      new boost::thread(boost::bind(&PointcloudDisplay::spin, this));
+}
+
+void PointcloudDisplay::stopSpin() {
+    this->continueFlag.clear(std::memory_order_relaxed);
+    this->viewer_thread->join();
+}
+
+void PointcloudDisplay::spin() {
+    // Initialize viewer
+    this->viewer = boost::shared_ptr<pcl::visualization::PCLVisualizer>(
+      new pcl::visualization::PCLVisualizer("Display"));
+    this->viewer->initCameraParameters();
+    this->viewer->setBackgroundColor(0, 0, 0);
+    this->viewer->addCoordinateSystem(1.0);
+    while (this->continueFlag.test_and_set(std::memory_order_relaxed) &&
+           !(this->viewer->wasStopped())) {
+        this->viewer->spinOnce(1);
+        boost::this_thread::sleep(boost::posix_time::microseconds(1000));
+
+        boost::mutex::scoped_lock cld_lock(this->update_mutex);
+        this->updateInternal();
+        cld_lock.unlock();
+    }
+    // Cleanup viewer
+    this->viewer->close();
+    this->viewer.reset();
+}
+
+void PointcloudDisplay::addPointcloud(const PCLPointCloud cld, int id) {
+    boost::mutex::scoped_lock lock(this->update_mutex);
+    this->clouds.emplace(Cloud{cld, id});
+    lock.unlock();
+}
+
+void PointcloudDisplay::addLine(pcl::PointXYZ pt1,
+                                pcl::PointXYZ pt2,
+                                int id1,
+                                int id2) {
+    boost::mutex::scoped_lock lock(this->update_mutex);
+    this->lines.emplace(Line{pt1, pt2, id1, id2});
+    lock.unlock();
+}
+
+void PointcloudDisplay::updateInternal() {
+    // add or update clouds in the viewer until the queue is empty
+    while (this->clouds.size() != 0) {
+        if (this->viewer->contains(std::to_string(this->clouds.front().id))) {
```

Comment:
You used `this->clouds.front().id` a couple of times, make a tmp variable above, same with `this->lines.front()` below.